lib/clients/public.js

```diff
@@ -1,206 +1,213 @@
-    var self = this;
+  getProductTicker(callback) {
+    return this.get(['products', this.productID, 'ticker'], callback);
+  }
 
-    args = args || {}
-    if (!callback && (typeof args === 'function')) {
+  getProductTrades(args = {}, callback) {
+    if (!callback && typeof args === 'function') {
       callback = args;
       args = {};
     }
+    return this.get(
+      ['products', this.productID, 'trades'],
+      { qs: args },
+      callback
+    );
+  }
 
-    var opts = {'qs': args};
-    return prototype.get.call(self, ['products', self.productID, 'trades'], opts, callback);
-  };
-
-  prototype.getProductTradeStream = function(tradesFrom, tradesTo) {
-    var self = this;
-    var shouldStop = null;
+  getProductTradeStream(tradesFrom, tradesTo) {
+    let shouldStop = null;
 
     if (typeof tradesTo === 'function') {
       shouldStop = tradesTo;
       tradesTo = null;
     }
 
-    var rs = new Readable({objectMode: true});
-    var started = false;
+    const rs = new Readable({ objectMode: true });
+    let started = false;
 
-    rs._read = function() {
+    rs._read = () => {
       if (!started) {
         started = true;
-        fetchTrades(self, rs, tradesFrom, tradesTo, shouldStop, 0);
+        fetchTrades.bind(this)(rs, tradesFrom, tradesTo, shouldStop, 0);
       }
     };
 
     return rs;
-  }
 
-  function fetchTrades(ctx, stream, tradesFrom, tradesTo, shouldStop) {
-    var after = tradesFrom + API_LIMIT + 1;
-    var loop = true;
+    function fetchTrades(stream, tradesFrom, tradesTo, shouldStop) {
+      let after = tradesFrom + this.API_LIMIT + 1;
+      let loop = true;
 
-    if (tradesTo && tradesTo <= after) {
-      after = tradesTo;
-      loop = false;
-    }
-
-    var opts = { before: tradesFrom, after: after, limit: API_LIMIT };
-
-    prototype.getProductTrades.call(ctx, opts, function(err, resp, data) {
-      if (err) {
-        stream.emit('error', err);
-        return;
+      if (tradesTo && tradesTo <= after) {
+        after = tradesTo;
+        loop = false;
       }
 
-      if (resp.statusCode === 429) {
-        // rate-limited, try again
-        setTimeout(function() {
-          fetchTrades(ctx, stream, tradesFrom, tradesTo, shouldStop);
-        }, 900);
-        return;
-      }
+      let opts = { before: tradesFrom, after: after, limit: this.API_LIMIT };
 
-      if (resp.statusCode !== 200) {
-        stream.emit('error', new Error('Encountered status code ' + resp.statusCode));
-      }
+      this.getProductTrades(opts, (err, resp, data) => {
+        if (err) {
+          stream.emit('error', err);
+          return;
+        }
 
-      for (var i = data.length - 1; i >= 0; i--) {
-        if (shouldStop && shouldStop(data[i])) {
-          stream.push(null);
+        if (resp.statusCode === 429) {
+          // rate-limited, try again
+          setTimeout(() => {
+            fetchTrades.call(this, stream, tradesFrom, tradesTo, shouldStop);
+          }, 900);
```

Comment:
The rate limit is per second, this skips to the next second. Not sure if this is the ideal solution, we could also just trigger a couple more requests.